### ISSUE TITLE
Memory blowup when parsing a big qasm file

### ISSUE THREAD

**Issue by eddieschoute** [Qiskit/qiskit-terra#1168]
### Informations

- **Qiskit Terra version**: qiskit 0.5.6
- **Python version**: python 3.7
- **Operating system**: MacOS 10.14

### What is the current behavior?
When reading in a (large) qasm file of about 40 megabytes in size to a DAGCircuit, the memory requirements grow to many gigabytes. I just tested that it grew up to 4 GB, but I think it will continue unrolling until an approximated 8-10GB is used.

### Steps to reproduce the problem
Run the following script with the attached qasm file in the same directory.
```
from pathlib import Path
from time import sleep
from qiskit.qasm import Qasm
from qiskit.unroll import Unroller, DAGBackend

BASIS = ["u1", "u2", "u3", "cx"]

with Path("highmem.qasm").open(mode='r') as randomfile:
    circuit_str = randomfile.read()
    program_node_circuit = Qasm(data=circuit_str).parse()
    circuit = Unroller(program_node_circuit, DAGBackend(basis=BASIS)).execute()

    print("Done reading")
    sleep(10)
```

[highmem.qasm.zip](https://github.com/Qiskit/qiskit-terra/files/2530984/highmem.qasm.zip)

### What is the expected behavior?
I would expect a memory use on the same order of magnitude of the qasm file.

### Suggested solutions
Profile the memory of datastructures and find the culprit.

**Comment by Jaxo** [comment]
For future reference...

Simply reproduce with:
<pre>
from qiskit.qasm import Qasm
Qasm(filename=“highmem.qasm”).parse()
</pre>

and qasm file:
<pre>
OPENQASM 2.0;
qreg q[81];
CX q[58],q[54];
CX q[49],q[45];
CX q[37],q[36];
CX q[33],q[27];
CX q[26],q[18];
CX q[11],q[9];
CX q[18],q[9];
CX q[25],q[18];
CX q[11],q[9];
CX q[45],q[9];
(the original input file has >2M lines of these)
</pre>

Here's a diff heap dump of each of those lines being parsed in `yacc.py::parseopt_notrack()`. 

<pre>
                                           types |   # objects |   total size
================================================ | =========== | ============
                                     class  dict |           7 |    944     B
                                     class  list |           7 |    496     B
                 class  qiskit.qasm._node._id.Id |           2 |    112     B
           class  qiskit.qasm._node._intnode.Int |           2 |    112     B
   class  qiskit.qasm._node._indexedid.IndexedId |           2 |    112     B
             class  qiskit.qasm._node._cnot.Cnot |           1 |     56     B
</pre>

I haven't gotten to the point of figuring out if that's an inefficiency in the parser (~2K per line!) or a memory issue with Python/PLY (a leak or fragmentation in the heap).

Here's the memory analyzer snippet:

<pre>
        from pympler import muppy, summary
        sum1 = summary.summarize(muppy.get_objects())
        ...parse a line...
        sum2 = summary.summarize(muppy.get_objects())
        summary.print_(summary.get_diff(sum1, sum2))
</pre>

**Comment by latone** [comment]
For future reference...

Simply reproduce with:
<pre>
from qiskit.qasm import Qasm
Qasm(filename=“highmem.qasm”).parse()
</pre>

and qasm file:
<pre>
OPENQASM 2.0;
qreg q[81];
CX q[58],q[54];
CX q[49],q[45];
CX q[37],q[36];
CX q[33],q[27];
CX q[26],q[18];
CX q[11],q[9];
CX q[18],q[9];
CX q[25],q[18];
CX q[11],q[9];
CX q[45],q[9];
(the original input file has >2M lines of these)
</pre>

Here's a diff heap dump of each of those lines being parsed in `yacc.py::parseopt_notrack()`. 

<pre>
                                           types |   # objects |   total size
================================================ | =========== | ============
                                     class  dict |           7 |    944     B
                                     class  list |           7 |    496     B
                 class  qiskit.qasm._node._id.Id |           2 |    112     B
           class  qiskit.qasm._node._intnode.Int |           2 |    112     B
   class  qiskit.qasm._node._indexedid.IndexedId |           2 |    112     B
             class  qiskit.qasm._node._cnot.Cnot |           1 |     56     B
</pre>

I haven't gotten to the point of figuring out if that's an inefficiency in the parser (~2K per line!) or a memory issue with Python/PLY (a leak or fragmentation in the heap).

Here's the memory analyzer snippet:

<pre>
        from pympler import muppy, summary
        sum1 = summary.summarize(muppy.get_objects())
        ...parse a line...
        sum2 = summary.summarize(muppy.get_objects())
        summary.print_(summary.get_diff(sum1, sum2))
</pre>

**Comment by eddieschoute** [comment]
Is there suggested approach to solve this problem yet? Maybe find the bug in parser generator or rewrite the parser?
By the way, it may be nicer to create a standalone project in the context of OpenQasm that handles parsing and generating OpenQasm files. To reduce coupling and provide a clean interface for people who would like to interoperate with OpenQasm.

**Comment by jlatone** [comment]
Eddie, I tried my simple 2-line test (from https://github.com/Qiskit/qiskit-terra/issues/1168#issuecomment-440788841) on the latest Terra, both 0.7.0 and 0.8.0. They both performed the same, but I think something may have been improved from pre-0.7.0. There's still constant linear growth during parse() but perhaps not as steep. Here's the `psrecord ` output from running your original 2.25M line file (runtime was 405sec consuming 5.5GB, not ideal but usable for now).

![plotbig](https://user-images.githubusercontent.com/676063/51946911-e8ab6c00-23d7-11e9-9cd6-a46a255df1e4.png)

```
# Elapsed time   CPU (%)     Real (MB)   Virtual (MB)
       0.000        0.000        5.281     4154.238
       1.001      152.900      197.195     4719.770
       2.001       99.600      219.074     4741.270
       3.003       99.900      239.621     4762.051
       4.006       99.500      259.172     4781.391
       5.009       99.900      278.129     4800.449
       6.013       99.700      293.188     4815.418
       7.018       99.900      306.027     4827.918
       8.022       99.900      335.082     4856.980
       9.026       99.700      346.676     4868.730
      10.030       99.900      373.188     4895.230
      11.035       99.900      378.273     4900.914
      12.039       99.300      406.754     4929.164
      13.043       99.600      419.527     4941.414
      14.048      100.100      432.074     4953.914
      15.051       99.800      458.750     4980.664
      16.053       99.700      478.504     5000.414
      17.058       99.700      478.504     5000.414
      18.063       99.900      498.828     5020.789
      19.068       99.900      529.508     5051.289
      20.073       99.800      551.926     5073.789
      21.078       99.900      551.926     5073.789
      22.079       99.900      562.965     5086.070
      23.084       99.700      593.180     5116.320
      24.086       99.500      620.992     5144.234
      25.087       99.800      643.508     5165.484
      26.092       99.900      643.508     5165.484
      27.093       99.900      647.129     5168.984
      28.098       99.600      674.555     5198.102
      29.099       99.100      701.859     5225.102
      30.100       99.800      730.117     5253.805
      31.100       99.900      757.238     5280.805
      32.105       99.200      758.863     5280.805
      33.108      100.300      758.863     5280.805
      34.113       99.000      766.691     5288.555
      35.117       99.900      795.055     5319.031
      36.118       99.800      824.188     5348.031
      37.123       99.900      854.359     5378.539
      38.126       99.700      882.957     5407.039
      39.131       99.300      901.004     5425.039
      40.136      100.400      903.059     5425.039
      41.140       99.900      903.059     5425.039
      42.144       98.800      903.059     5425.039
      43.149      100.700      924.641     5446.539
      44.152       99.800      951.699     5476.082
      45.153       99.900      981.805     5506.082
      46.154       99.700     1010.164     5534.332
      47.158       99.900     1040.172     5564.906
      48.159       99.700     1068.691     5593.406
      49.160       99.600     1082.551     5604.656
      50.163       99.900     1082.551     5604.656
      51.168      100.300     1082.551     5604.656
      52.169       99.600     1082.551     5604.656
      53.171      100.200     1091.578     5613.656
      54.171       99.800     1120.828     5642.656
      55.172       99.900     1148.223     5673.273
      56.178       99.800     1177.773     5702.898
      57.183       99.900     1208.066     5733.148
      58.184       99.800     1236.773     5761.898
      59.186      100.000     1266.848     5792.559
      60.188       99.900     1295.754     5821.309
      61.192       99.900     1303.668     5829.309
      62.193       99.800     1306.965     5829.309
      63.194       99.800     1306.965     5829.309
      64.195       99.100     1306.965     5829.309
      65.199      100.200     1306.965     5829.309
      66.205      100.500     1313.918     5836.059
      67.208       99.900     1344.402     5866.559
      68.211       99.700     1372.707     5894.809
      69.215       99.900     1398.906     5925.023
      70.217       99.800     1427.840     5953.773
      71.219      100.000     1458.148     5984.023
      72.222       99.600     1487.000     6012.773
      73.226       99.900     1517.652     6043.520
      74.232       99.900     1546.457     6073.043
      75.235       99.600     1576.520     6103.043
      76.240       99.800     1583.570     6110.043
      77.244      100.000     1587.742     6110.043
      78.250       98.800     1587.742     6110.043
      79.255      100.000     1587.742     6110.043
      80.260      100.700     1587.742     6110.043
      81.262       99.800     1587.742     6110.043
      82.264      100.100     1589.477     6111.793
      83.265       99.800     1620.223     6142.543
      84.266       99.800     1649.414     6171.543
      85.268      100.000     1680.203     6202.293
      86.269       99.800     1704.750     6232.129
      87.270       99.900     1735.023     6262.129
      88.276       99.700     1764.184     6291.379
      89.280       99.900     1794.449     6321.379
      90.282       99.800     1823.172     6350.129
      91.288       99.900     1853.383     6380.129
      92.292       99.800     1881.508     6409.539
      93.293       99.900     1911.941     6439.789
      94.296       99.200     1932.402     6460.289
      95.297      100.100     1937.684     6460.289
      96.298       99.700     1937.684     6460.289
      97.303      100.200     1937.684     6460.289
      98.306       99.900     1937.684     6460.289
      99.310       99.900     1937.684     6460.289
     100.314       99.500     1937.684     6460.289
     101.319      100.200     1937.684     6460.289
     102.321       99.600     1937.684     6460.289
     103.326       99.800     1961.703     6484.039
     104.327       99.900     1991.547     6513.789
     105.332       99.900     2022.418     6544.789
     106.336       99.800     2051.340     6573.539
     107.341       99.800     2081.691     6603.789
     108.346       99.900     2104.820     6633.531
     109.348       99.900     2135.418     6664.031
     110.353       99.800     2164.609     6693.031
     111.357       99.900     2195.379     6723.781
     112.360       99.800     2224.703     6753.281
     113.365       99.900     2255.660     6784.031
     114.370       99.800     2285.031     6813.281
     115.373       99.900     2315.461     6843.781
     116.378       99.800     2344.070     6873.867
     117.384       99.300     2369.070     6898.617
     118.389       99.900     2369.070     6898.617
     119.392       99.700     2375.754     6898.617
     120.394       99.800     2375.754     6898.617
     121.399      100.500     2375.754     6898.617
     122.399       99.600     2375.754     6898.617
     123.404       99.900     2375.754     6898.617
     124.406       99.300     2375.754     6898.617
     125.411       99.800     2375.754     6898.617
     126.412       99.700     2375.754     6898.617
     127.417      100.000     2375.754     6898.617
     128.419      100.400     2394.789     6917.617
     129.420      100.000     2425.801     6948.617
     130.422       99.900     2455.008     6977.617
     131.427      100.000     2486.207     7008.867
     132.432       99.500     2515.441     7037.867
     133.438       99.900     2546.641     7069.117
     134.443       99.900     2576.262     7098.617
     135.445       99.900     2598.699     7129.309
     136.446       99.900     2627.934     7158.559
     137.450       99.800     2658.449     7188.809
     138.455       99.800     2687.965     7218.309
     139.459       99.900     2718.863     7249.059
     140.460       99.800     2747.988     7278.059
     141.464      100.000     2778.895     7309.059
     142.469       99.600     2807.500     7337.559
     143.470       99.900     2838.430     7368.309
     144.472       99.800     2868.016     7397.809
     145.473      100.000     2898.375     7430.109
     146.478       99.400     2915.082     7446.859
     147.481      100.200     2915.082     7446.859
     148.485       99.700     2923.543     7446.859
     149.487      100.000     2923.543     7446.859
     150.492       99.700     2923.543     7446.859
     151.496      100.000     2923.543     7446.859
     152.497       99.800     2923.543     7446.859
     153.500       99.400     2923.543     7446.859
     154.503      100.400     2923.543     7446.859
     155.504       99.900     2923.543     7446.859
     156.509       99.400     2923.543     7446.859
     157.514      100.400     2923.543     7446.859
     158.519       99.400     2923.543     7446.859
     159.523       99.800     2923.543     7446.859
     160.527      100.400     2942.477     7465.609
     161.531       99.900     2973.816     7496.859
     162.535       99.600     3003.172     7526.109
     163.541      100.000     3033.625     7556.359
     164.543       99.800     3063.129     7585.859
     165.544       99.900     3094.078     7616.859
     166.550       99.600     3122.457     7645.109
     167.551       99.800     3152.797     7675.359
     168.556       99.800     3182.477     7704.859
     169.560       99.900     3213.008     7735.359
     170.561       99.800     3232.812     7765.801
     171.561       99.900     3263.043     7795.801
     172.565       99.800     3292.555     7825.301
     173.567       99.800     3322.723     7855.301
     174.572       99.900     3352.082     7884.551
     175.573       99.700     3345.719     7911.801
     176.574       99.300     3363.469     7939.551
     177.574       99.700     3369.730     7969.301
     178.580       99.900     3353.273     7998.551
     179.580       99.900     3334.559     8028.051
     180.584       99.900     3343.418     8057.551
     181.585       99.800     3335.484     8086.551
     182.586       99.600     3317.766     8115.301
     183.589       99.800     3324.984     8128.801
     184.593       99.900     3325.363     8128.801
     185.594       97.100     3378.512     8128.801
     186.598       99.500     3475.840     8128.801
     187.602       99.600     3486.445     8128.801
     188.604      100.300     3486.445     8128.801
     189.604      100.000     3598.945     8128.801
     190.606       99.400     3598.945     8128.801
     191.610      100.400     3598.945     8128.801
     192.615       99.500     3598.945     8128.801
     193.619      100.100     3598.945     8128.801
     194.625       99.600     3598.945     8128.801
     195.629      100.200     3598.945     8128.801
     196.631       99.900     3598.945     8128.801
     197.636       99.800     3598.945     8128.801
     198.639       99.800     3598.945     8128.801
     199.640       99.700     3598.945     8128.801
     200.641      100.000     3598.945     8128.801
     201.646       99.900     3617.000     8158.891
     202.646       99.800     3646.973     8188.891
     203.647       99.900     3677.387     8219.141
     204.649       99.900     3707.555     8249.141
     205.650       99.900     3737.453     8278.891
     206.656       99.900     3767.730     8309.141
     207.660       99.800     3797.652     8339.141
     208.661       99.900     3827.641     8368.891
     209.665       99.900     3856.926     8398.641
     210.669       99.800     3886.168     8428.141
     211.669       99.800     3884.730     8457.141
     212.674       99.500     3855.473     8486.391
     213.674       99.900     3860.789     8515.141
     214.675       99.800     3890.914     8544.891
     215.678       99.900     3919.828     8576.395
     216.683       99.900     3950.133     8606.395
     217.683       99.600     3979.754     8635.645
     218.688       99.900     4009.406     8665.145
     219.688       99.900     4022.199     8694.395
     220.693       99.900     4002.773     8724.145
     221.695       99.700     3984.555     8753.145
     222.700       99.700     3967.492     8782.145
     223.701       99.800     3913.680     8812.395
     224.705       99.900     3943.730     8842.145
     225.710       99.800     3973.383     8871.645
     226.711       99.900     4003.508     8901.395
     227.712       99.800     4032.035     8930.395
     228.714       99.900     4061.867     8959.895
     229.715       99.600     4082.977     8984.145
     230.715      100.200     4083.176     8984.145
     231.718       99.200     4083.203     8984.145
     232.720       98.900     4276.066     8984.145
     233.720       99.700     4276.371     8984.145
     234.723      100.100     4276.383     8984.145
     235.723       90.500     4273.711     8984.145
     236.729       99.600     4344.055     8984.145
     237.729       99.600     4344.055     8984.145
     238.734      100.100     4344.055     8984.145
     239.735       90.700     4344.055     8984.145
     240.741       99.700     4344.055     8984.145
     241.744       99.700     4344.055     8984.145
     242.744       99.600     4344.055     8984.145
     243.745      100.000     4344.055     8984.145
     244.745       99.100     4344.055     8984.145
     245.746      100.700     4344.055     8984.145
     246.748       99.900     4344.055     8984.145
     247.749       99.700     4344.055     8984.145
     248.751       99.500     4344.055     8984.145
     249.752      100.000     4344.051     8984.145
     250.757       99.100     4344.043     8984.145
     251.761       99.700     4344.160     8984.145
     252.762       99.800     4374.051     9013.895
     253.762       99.700     4403.285     9042.645
     254.764       99.900     4420.621     9075.090
     255.765       99.800     4449.941     9104.090
     256.770      100.000     4481.113     9134.840
     257.773       99.500     4508.391     9163.090
     258.776       99.800     4538.438     9192.840
     259.777       99.700     4567.898     9221.840
     260.782       99.900     4598.641     9252.340
     261.786       99.800     4628.441     9281.840
     262.788       99.600     4658.855     9311.840
     263.788       99.800     4688.117     9340.840
     264.791       99.800     4718.805     9371.340
     265.794       99.000     4747.242     9399.340
     266.797       98.700     4773.145     9425.090
     267.799       98.000     4797.961     9449.590
     268.799       99.700     4825.934     9477.340
     269.799       99.400     4851.465     9502.590
     270.803       99.500     4878.148     9529.090
     271.809       99.700     4906.418     9557.090
     272.811       99.500     4933.320     9583.590
     273.812       99.200     4957.844     9611.496
     274.814       99.600     4983.973     9637.246
     275.814       99.400     5009.660     9662.746
     276.815       99.900     5039.883     9692.746
     277.821       99.800     5069.246     9721.746
     278.826       99.900     5098.543     9750.746
     279.826       98.500     5122.512     9774.496
     280.829       99.400     5147.551     9799.246
     281.830       98.900     5170.926     9822.496
     282.835       99.400     5195.477     9846.746
     283.835       98.600     5218.746     9869.746
     284.840       99.300     5243.941     9894.746
     285.844       98.100     5264.395     9914.996
     286.845       99.200     5268.613     9937.246
     287.848       97.700     5270.816     9959.496
     288.848       99.500     5291.051     9984.996
     289.850       98.300     5275.012    10008.746
     290.853       99.600     5280.551    10034.996
     291.854       98.900     5286.488    10051.246
     292.854       99.100     5281.715    10051.246
     293.857       99.600     5274.188    10051.246
     294.861       99.600     5272.688    10051.246
     295.862       96.800     5376.871    10051.246
     296.863      100.200     5377.020    10051.246
     297.863       99.000     5377.039    10051.246
     298.864      100.400     5377.039    10051.246
     299.864       99.200     5377.039    10051.246
     300.869       99.700     5417.922    10051.246
     301.873       99.100     5417.922    10051.246
     302.875       99.700     5417.922    10051.246
     303.879       98.800     5417.922    10051.246
     304.879       99.700     5417.922    10051.246
     305.880       99.900     5417.922    10051.246
     306.884       99.200     5417.922    10051.246
     307.888      100.000     5417.922    10051.246
     308.889       99.500     5417.922    10051.246
     309.894       99.000     5417.922    10051.246
     310.897       99.700     5417.922    10051.246
     311.899       97.100     5417.922    10051.246
     312.899       99.600     5417.922    10051.246
     313.903       98.900     5417.922    10051.246
     314.904       99.400     5417.922    10051.246
     315.909       99.700     5417.922    10051.246
     316.913       99.200     5417.922    10051.246
     317.918       99.500     5417.922    10051.246
     318.923       99.200     5417.922    10051.246
     319.928       98.600     5417.922    10051.246
     320.934       99.000     5421.281    10054.746
     321.935       99.600     5449.621    10082.746
     322.939       98.800     5471.812    10109.496
     323.940       99.600     5501.168    10138.488
     324.944       99.700     5529.512    10166.738
     325.949       99.200     5554.996    10192.738
     326.953       99.600     5506.863    10204.332
     327.958       98.900     5502.383    10208.082
     328.959       99.400     5495.879    10208.082
     329.960      100.100     5496.770    10208.082
     330.961       98.000     5484.180    10208.082
     331.962       99.100     5563.688    10208.082
     332.964       99.200     5567.836    10208.082
     333.965       98.800     5581.355    10208.082
     334.966      100.200     5581.355    10208.082
     335.966       99.100     5581.355    10208.082
     336.969      100.600     5590.344    10208.082
     337.974       99.300     5593.441    10208.082
     338.981      100.300     5593.441    10208.082
     339.984       98.600     5593.441    10208.082
     340.988      100.600     5593.441    10208.082
     341.990       98.100     5593.441    10208.082
     342.995       99.700     5593.441    10208.082
     343.997      100.600     5593.441    10208.082
     345.004      100.100     5593.441    10208.082
     346.006       99.800     5593.441    10208.082
     347.008       98.800     5593.441    10208.082
     348.010      100.700     5593.441    10208.082
     349.012       99.800     5593.441    10208.082
     350.017      100.000     5593.441    10208.082
     351.018       99.400     5593.441    10208.082
     352.019       99.700     5593.441    10208.082
     353.024      100.600     5593.441    10208.082
     354.025       99.500     5593.441    10208.082
     355.026       99.900     5593.441    10208.082
     356.032      100.100     5593.441    10208.082
     357.034       99.400     5593.441    10208.082
     358.039      100.100     5593.820    10208.082
     359.040       99.900     5593.820    10208.082
     360.042       99.800     5593.820    10208.082
     361.048      100.000     5593.820    10208.082
     362.049       99.100     5593.820    10208.082
     363.050       99.500     5593.820    10208.082
     364.051       99.800     5593.820    10208.082
     365.053      100.200     5593.820    10208.082
     366.058       99.700     5593.820    10208.082
     367.059       95.400     5582.520    10208.082
     368.060       95.200     5577.059    10208.082
     369.065       99.000     5574.148    10208.082
     370.070      100.500     5575.219    10208.082
     371.076       99.400     5575.137    10208.082
     372.076       98.300     5575.164    10208.082
     373.077      100.100     5575.164    10208.082
     374.082       99.200     5575.164    10208.082
     375.083       99.600     5575.164    10208.082
     376.084      100.500     5575.164    10208.082
     377.086       99.200     5575.164    10208.082
     378.087      100.100     5575.164    10208.082
     379.088      100.100     5575.164    10208.082
     380.089       99.600     5575.164    10208.082
     381.090       99.900     5575.164    10208.082
     382.091       99.500     5575.164    10208.082
     383.093       99.700     5575.164    10208.082
     384.094      100.300     5575.164    10208.082
     385.095       99.700     5575.180    10208.082
     386.101      100.600     5575.180    10208.082
     387.102       99.900     5575.180    10208.082
     388.103       99.200     5575.180    10208.082
     389.107      100.400     5575.180    10208.082
     390.109       99.700     5575.180    10208.082
     391.111       99.300     5575.180    10208.082
     392.112       99.800     5575.180    10208.082
     393.113      100.300     5581.211    10208.082
     394.114      100.000     5582.168    10208.082
     395.115       99.900     5582.391    10208.082
     396.116       99.300     5582.391    10208.082
     397.117       99.900     5582.391    10208.082
     398.122       99.800     5582.391    10208.082
     399.123      100.400     5582.391    10208.082
     400.124       99.700     5582.391    10208.082
     401.125       99.500     5582.391    10208.082
     402.127      100.100     5582.391    10208.082
     403.128      100.000     4417.008     8967.703
     404.131       99.900     2634.793     7185.203
     405.136      100.000      846.871     5396.828


```

**Comment by jlatone** [comment]
And, Eddie, as for "By the way, it may be nicer...", I think you should enter a separate ("good first") Issue to request an Enhancement so that you idea is not lost here.

**Comment by eddieschoute** [comment]
The performance may be slightly improved, but requiring 5.5GB of memory for a 2.25MB file (a 2000x blowup) is still far from ideal.

**Comment by 1ucian0** [comment]
I renamed the issue to reflect that is a problem of the parser. Feel free to roll that back if you disagree.

**Comment by chetmurthy** [comment]
I did a little investigation.  I took "highmem.qasm" and used "head -NN" to produce prefixes of the file of length 2^N, for N in [10...21].  Comparing the memory used for (e.g.) 2^20, 2^21 lines, and other pairs, I found that the per-line cost is about 2.6k bytes.  And this doesn't change much over this data-series.

So there's not some "bug", I fear.  Just really, really wasteful code.  Maybe it's fixable, maybe it isn't.  I'll keep digging, but figured I'd report it.

**Comment by chetmurthy** [comment]
TL;DR the wastefulness here is due to Python, and there's not much that can be done to fix it.

OK, I'm *pretty sure* that this isn't a bug in the Qasm Python code, nor in PLY (Python parser/lexer generator).  Here's a small test:
```
qasm_string = """
CX q[0],q[1];
"""

def main():
    q = QasmParser(None)
    ast = q.parse(data=qasm_string)
    cnot = ast.children[0]
    c0 = cnot.children[0]
    c1 = cnot.children[1]
    c0c0 = c0.children[0]
    pp = pprint.PrettyPrinter(indent=2)
    pp.pprint(c0c0.to_ppable())
    count = int(sys.argv[1])
    l = []
    for i in range(1,count+1):
        l.append(copy.deepcopy(c0c0))
```
[N.B.  I disabled the declared-variable check ("verify_reg") in the parser, for this test]
See that line "deepcopy(c0c0)"?  It started as "deepcopy(cnot)", then "deepcopy(c0)".

This program takes an integer parameter, and appends that many deep-copies of the selected node onto a list.  Running this with (524288, 1048576), on "cnot", I find that each "cnot" line consumes about 1.8k of memory.  Doing it on "c0c0" (the "q" in the QASM program) goes to "0.278k".  So then I replaced with
```
def main():
    id = Id("q", 1, "foo.qasm")
    pp = pprint.PrettyPrinter(indent=2)
    pp.pprint(id.to_ppable())
    count = int(sys.argv[1])
    l = []
    for i in range(1,count+1):
        l.append(Id("q", 1, "foo.qasm"))
```
And this has the same memory-cost as the "c0c0" case above.  Which tells us that even though we're not doing any lexing/parsing, the per-node cost is unchanged (278 bytes per "id" node).

There's not that much "beneath" the "Id" class.  I doubt that there's anything to be done about this. Python wasn't meant for complex programs, after all.

We CAN do something to improve on this.  For instance, we could "hash-cons" all identifiers and constants, which would result in some memory-savings.  We could hash-cons all expressions.  If the AST is never (NEVER) modified by downstream code, this would be safe.  I performed a simple experiment to test this.  In the first example above (with the "deepcopy(cnot)") we had "Cnot([c0, c1])" (so, construct a new "Cnot" node each iteration, but reuse its child nodes), the cost of each Cnot comes down to 253 bytes (from 1.8k.  This shouldn't be surprising.  But this also won't give any sort of dramatic savings.  Just a small constant-factor.
